src/hotspot/share/gc/shenandoah/shenandoahUncommitThread.cpp

```diff
@@ -0,0 +1,200 @@
+
+  if (_heap->committed() <= shrink_until) {
+    return false;
+  }
+
+  for (size_t i = 0; i < _heap->num_regions(); i++) {
+    ShenandoahHeapRegion *r = _heap->get_region(i);
+    if (r->is_empty_committed() && (r->empty_time() < shrink_before)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void ShenandoahUncommitThread::notify_soft_max_changed() {
+  assert(is_uncommit_allowed(), "Only notify if uncommit is allowed");
+  if (_soft_max_changed.try_set()) {
+    MonitorLocker locker(&_stop_lock, Mutex::_no_safepoint_check_flag);
+    locker.notify_all();
+  }
+}
+
+void ShenandoahUncommitThread::notify_explicit_gc_requested() {
+  assert(is_uncommit_allowed(), "Only notify if uncommit is allowed");
+  if (_explicit_gc_requested.try_set()) {
+    MonitorLocker locker(&_stop_lock, Mutex::_no_safepoint_check_flag);
+    locker.notify_all();
+  }
+}
+
+bool ShenandoahUncommitThread::is_uncommit_allowed() const {
+  return _uncommit_allowed.is_set();
+}
+
+void ShenandoahUncommitThread::uncommit(double shrink_before, size_t shrink_until) {
+  assert(ShenandoahUncommit, "should be enabled");
+  assert(_uncommit_in_progress.is_unset(), "Uncommit should not be in progress");
+
+  if (!is_uncommit_allowed()) {
+    return;
+  }
+
+  const char* msg = "Concurrent uncommit";
+  EventMark em("%s", msg);
+  double start = os::elapsedTime();
+  size_t committed_start = _heap->committed();
+  log_info(gc, start)("%s", msg);
+
+  _uncommit_in_progress.set();
+
+  // Application allocates from the beginning of the heap, and GC allocates at
+  // the end of it. It is more efficient to uncommit from the end, so that applications
+  // could enjoy the near committed regions. GC allocations are much less frequent,
+  // and therefore can accept the committing costs.
+  size_t count = 0;
+  for (size_t i = _heap->num_regions(); i > 0; i--) {
+    if (!is_uncommit_allowed()) {
+      break;
+    }
+
+    ShenandoahHeapRegion* r = _heap->get_region(i - 1);
+    if (r->is_empty_committed() && (r->empty_time() < shrink_before)) {
+      SuspendibleThreadSetJoiner sts_joiner;
+      ShenandoahHeapLocker locker(_heap->lock());
+      if (r->is_empty_committed()) {
+        if (_heap->committed() < shrink_until + ShenandoahHeapRegion::region_size_bytes()) {
+          break;
+        }
+
+        r->make_uncommitted();
+        count++;
+      }
+    }
+    SpinPause(); // allow allocators to take the lock
+  }
+
+  {
+    MonitorLocker locker(&_uncommit_lock, Mutex::_no_safepoint_check_flag);
+    _uncommit_in_progress.unset();
+    locker.notify_all();
+  }
+
+  if (count > 0) {
+    _heap->notify_heap_changed();
+  }
+
+  size_t committed_end = _heap->committed();
+  double elapsed = os::elapsedTime() - start;
+  log_info(gc)("%s " PROPERFMT "(" PROPERFMT ") %.3fms",
+               msg, PROPERFMTARGS(committed_start - committed_end), PROPERFMTARGS(_heap->capacity()),
+               elapsed * MILLIUNITS);
```

Comment:
I think we want an additional space. I see the current output is:

```
[11.366s][info][gc] Concurrent uncommit 32768K(192M) 2.506ms
```

Should probably be:

```
[11.366s][info][gc] Concurrent uncommit 32768K (192M) 2.506ms
```